# Case 0. Learning basics
**Applications of Neural Networks in Medicine**<br>
21.10.2024<br>
Miska Nurminen<br>
[Information Technology, Bachelor's Degree](https://www.metropolia.fi/en/academics/bachelors-degrees/information-technology)<br>
[Metropolia University of Applied Sciences](https://www.metropolia.fi/en)

- **v3**: Simplified version based on discussion with JK.
- **v4**: Added conversion of labels [to_categorical](https://www.tensorflow.org/api_docs/python/tf/keras/utils/to_categorical) and changed the loss function to [categorical crossentropy](https://www.tensorflow.org/api_docs/python/tf/keras/metrics/categorical_crossentropy).
- **v5**: Changes in instructions wordings.
- **v6**: Small changes in model parameters.

## 1. Introduction

This Notebook was created for the purpose of understanding and learning the basics of neural networks. <br>
The main objectives were to find the most simple neural network with as few epochs as possible to achieve accuracy of 97%.

## 2. Setup

The only library used is Tensorflow. That is because none of this code would work without it.

In [287]:
import tensorflow as tf
print(f'tensorflow: {tf.__version__}')

tensorflow: 2.17.0


## 3. Dataset

The dataset consists of 60,000 28x28 grayscale images of the 10 digits, along with a test set of 10,000 images.

In [288]:
mnist = tf.keras.datasets.mnist

(x_train, y_train), (x_test, y_test) = mnist.load_data()

In [289]:
print(f'x_train: shape {x_train.shape} and ndim {x_train.ndim}')
print(f'x_test:  shape {x_test.shape} and ndim {x_test.ndim}')

print(f'y_train: shape {y_train.shape} and ndim {y_train.ndim}')
print(f'y_test:  shape {y_test.shape} and ndim {y_test.ndim}')

x_train: shape (60000, 28, 28) and ndim 3
x_test:  shape (10000, 28, 28) and ndim 3
y_train: shape (60000,) and ndim 1
y_test:  shape (10000,) and ndim 1


## 4. Preprocessing

The pixel values of the images range from 0 through 255. Scale these values to a range of 0 to 1 by dividing the values by 255.0.<br>
Train and test data is also converted to a binary class matrix from a class vector.

In [290]:
x_train = x_train / 255.0 
x_test = x_test / 255.0

y_train = tf.keras.utils.to_categorical(y_train)
y_test = tf.keras.utils.to_categorical(y_test)

## 5. Modeling

Sequential is useful for stacking layers where each layer has one input tensor and one output tensor.<br>
Layers are functions with a known mathematical structure that can be reused and have trainable variables.<br>
Most TensorFlow models are composed of layers.

In [291]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Flatten, Dense

model = Sequential([
    Flatten(input_shape=(28, 28)),
    Dense(80, activation='relu'),
    Dense(10, activation='softmax')
])

In [292]:
model.compile(optimizer = 'adam',
              loss = 'categorical_crossentropy',
              metrics = ['accuracy'])

In [293]:
model.summary()

Model: "sequential_32"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ flatten_32 (Flatten)            │ (None, 784)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_64 (Dense)                │ (None, 80)             │        62,800 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_65 (Dense)                │ (None, 10)             │           810 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 63,610 (248.48 KB)

 Trainable params: 63,610 (248.48 KB)

 Non-trainable params: 0 (0.00 B)

## 6. Training

Neural network training involves adjusting the model's weights based on input data to improve prediction accuracy.<br>
The network's output is compared to actual targets using a loss function.<br>
This process is repeated iteratively to optimize the model's performance.<br>
In this case, the process is repated 3 times.

In [294]:
model.fit(x_train, y_train, epochs = 3)

Epoch 1/3
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 2s 985us/step - accuracy: 0.8672 - loss: 0.4803
Epoch 2/3
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 2s 962us/step - accuracy: 0.9600 - loss: 0.1396
Epoch 3/3
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 2s 965us/step - accuracy: 0.9716 - loss: 0.0951


## 7. Performance and evaluation

The final perfomance of the model was good. It has an accuracy of 0.9713 and loss is 0.0985.

In [295]:
model.evaluate(x_test,  y_test, verbose=2)

313/313 - 0s - 1ms/step - accuracy: 0.9713 - loss: 0.0985


[0.09851877391338348, 0.9713000059127808]

## 8. Discussion and conclusions

- I played around with the amount of nodes in each of the 3 dense layers. In the end I achieved the wanted results with just 2 dense layers with 80 and 10 nodes respectively.
- The best model I got was a sequental model with 1 hidden layer of 80 nodes.
- The final perfomance of the model was good. It has an accuracy of 0.9713 and loss is 0.0985.
- I can't say about my observations but I learned the basics of neural networks and how they work.
- You could maybe use a different model to achieve better results.
- I also tried a CNN, but it was unnecessary. Accuracy was around 0.099.

To learn more, see [Tensorflow tutorials](https://www.tensorflow.org/tutorials).